src/xenia/apu/xma_context.cc

```diff
@@ -343,308 +333,458 @@ void XmaContext::DecodePackets(XMA_CONTEXT_DATA* data) {
-    int num_channels = data->is_stereo ? 2 : 1;
     if (!data->input_buffer_0_valid && !data->input_buffer_1_valid) {
       // Out of data.
       break;
     }
 
-    if (data->input_buffer_read_offset == 0) {
-      // Invalid offset. Go ahead and set it.
-      uint32_t offset = xma::GetPacketFrameOffset(current_input_buffer);
-      if (offset == -1) {
-        // No more frames.
-        if (data->current_buffer == 0) {
-          data->input_buffer_0_valid = 0;
-          data->input_buffer_read_offset = 0;
-          data->current_buffer++;
-        } else if (data->current_buffer == 1) {
-          data->input_buffer_1_valid = 0;
-          data->input_buffer_read_offset = 0;
-          data->current_buffer--;
-        }
+    // assert_true(packets_skip_ == 0);
+    // assert_true(split_frame_len_ == 0);
+    // assert_true(split_frame_len_partial_ == 0);
 
-        // Die if we have no partial saved.
-        if (!partial_frame_saved_) {
+    // Where are we in the buffer (in XMA jargon)
+    int packet_idx, frame_idx, frame_count;
+    uint8_t* packet;
+    bool frame_last_split;
+
+    BitStream stream(current_input_buffer, current_input_size * 8);
+    stream.SetOffset(data->input_buffer_read_offset);
+
+    // if we had a buffer swap try to skip packets first
+    if (packets_skip_ > 0) {
+      packet_idx =
+          GetFramePacketNumber(current_input_buffer, current_input_size,
+                               data->input_buffer_read_offset);
+      while (packets_skip_ > 0) {
+        packets_skip_--;
+        packet_idx++;
+        if (packet_idx >= current_input_packet_count) {
+          SwapInputBuffer(data);
           return;
         }
-      } else {
-        data->input_buffer_read_offset = offset;
       }
+      // invalid frame pointer but needed for us
+      data->input_buffer_read_offset = packet_idx * kBytesPerPacket * 8;
+      // continue;
     }
 
-    if (!ValidFrameOffset(current_input_buffer, current_input_size,
-                          data->input_buffer_read_offset)) {
-      XELOGAPU("XmaContext {}: Invalid read offset {}!", id(),
-               data->input_buffer_read_offset);
-      if (data->current_buffer == 0) {
-        data->current_buffer = 1;
-        data->input_buffer_0_valid = 0;
-      } else if (data->current_buffer == 1) {
-        data->current_buffer = 0;
-        data->input_buffer_1_valid = 0;
+    if (split_frame_len_) {
+      // handle a frame that was split over two packages
+      packet_idx =
+          GetFramePacketNumber(current_input_buffer, current_input_size,
+                               data->input_buffer_read_offset);
+      packet = current_input_buffer + packet_idx * kBytesPerPacket;
+      std::tie(frame_count, frame_last_split) = GetPacketFrameCount(packet);
+      frame_idx = -1;
+
+      stream = BitStream(current_input_buffer,
+                         ((packet_idx + 1) * kBytesPerPacket) * 8);
+      stream.SetOffset(packet_idx * kBytesPerPacket * 8 + 32);
```

Comment:
There's lots of `kBytesPerPacket * 8` in this code. Maybe add a new `kBitsPerPacket` constant?